File: app/src/main/java/org/sabda/gpt/Selengkapnya.java
package org.sabda.gpt;

import androidx.appcompat.app.AppCompatActivity;

import android.app.Dialog;
import android.content.Intent;
import android.os.Bundle;
import android.view.ViewGroup;
import android.view.Window;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.ImageView;

import com.github.chrisbanes.photoview.PhotoView;

import java.util.Objects;

public class Selengkapnya extends AppCompatActivity implements NetworkUtil.NetworkChangeCallback  {

    Button aidan, aisquared, fokus, coba, alkitabGPT;
    PhotoView Aidan, square, fokus1, AlkitabGPT, bibleGPT;
    Button v_aidan, p_aidan, v_alkitab, p_alkitab, v_aiSquared, p_aiSquared, v_fokus, p_fokus, v_bibleGpt, p_bibleGpt;
    ImageView back;
    boolean isConnected;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_selengkapnya);

        isConnected = NetworkUtil.isNetworkAvailable(this);
        NetworkUtil.registerNetworkChangeReceiver(this, this::updateConnectionStatus);

        initializeViews();
        updateConnectionStatus(isConnected);

        back.setOnClickListener(v -> finish());
        setImageResources();
        ClickListener();

    }

    private void initializeViews() {
        back = findViewById(R.id.kembali);

        /*inisialisasi button navigasi*/
        aidan = findViewById(R.id.aidanAlkitabButton);
        aisquared = findViewById(R.id.aisquaredbutton);
        fokus = findViewById(R.id.fokusbutton);
        alkitabGPT = findViewById(R.id.alkitabGPTbutton);
        coba = findViewById(R.id.cobabutton);

        /*inisialisasi button image*/
        Aidan = findViewById(R.id.aidanAlkitab);
        square = findViewById(R.id.AI_SquareSquare);
        fokus1 = findViewById(R.id.FOKUS);
        AlkitabGPT = findViewById(R.id.AlkitabGPT);
        bibleGPT = findViewById(R.id.bibleGpt);

        /*inisialisasi button video & pdf*/
        v_aidan = findViewById(R.id.videoaidanAlkitab);
        v_alkitab = findViewById(R.id.videoAlkitabGPT);
        v_aiSquared = findViewById(R.id.videoaiSquared);
        v_fokus = findViewById(R.id.videofokus);
        v_bibleGpt = findViewById(R.id.videoBibleGpt);

        p_aidan = findViewById(R.id.pdfaidanAlkitab);
        p_alkitab = findViewById(R.id.pdfAlkitabGPT);
        p_aiSquared = findViewById(R.id.pdfaiSquared);
        p_fokus = findViewById(R.id.pdffokus);
        p_bibleGpt = findViewById(R.id.pdfBibleGpt);
    }

    private void setImageResources() {
        Aidan.setImageResource(R.drawable.aidanalkitab);
        square.setImageResource(R.drawable.aisquare);
        fokus1.setImageResource(R.drawable.fokus);
        AlkitabGPT.setImageResource(R.drawable.alkitabgptdetail);
        bibleGPT.setImageResource(R.drawable.biblegpt);
    }

    private void ClickListener() {
        Aidan.setOnClickListener(v -> showImage(R.drawable.aidanalkitab));
        square.setOnClickListener(v -> showImage(R.drawable.aisquare));
        fokus1.setOnClickListener(v -> showImage(R.drawable.fokus));
        AlkitabGPT.setOnClickListener(v -> showImage(R.drawable.alkitabgptdetail));
        bibleGPT.setOnClickListener(v -> showImage(R.drawable.biblegpt));

        aidan.setOnClickListener(v -> handleClick("https://ai.sabda.org/ai-dan-alkitab/", "Seminar: AI dan Alkitab"));
        alkitabGPT.setOnClickListener(v -> handleClick("https://ai.sabda.org/alkitab-gpt/", "Seminar: Alkitab GPT"));
        aisquared.setOnClickListener(v -> handleClick("https://ai.sabda.org/pa-dengan-ai-cara-ai-squared/", "Metode AI Squared"));
        fokus.setOnClickListener(v -> handleClick("https://ai.sabda.org/bagaimana-ai/", "Metode F.0.K.U.S"));

        v_aidan.setOnClickListener(v -> handleYoutubeClick("gDLTyyabCvI", 218));
        v_alkitab.setOnClickListener(v -> handleYoutubeClick("Pf9DPEcX8JA", 797));
        v_aiSquared.setOnClickListener(v -> handleYoutubeClick("qHPwPjuAEs4", 783));
        v_fokus.setOnClickListener(v -> handleYoutubeClick("GHcRwNXdnvQ", 0));
        v_bibleGpt.setOnClickListener(v -> handleYoutubeClick("vZkGKgtRPeU", 0));

        p_aidan.setOnClickListener(v -> handlePdfClick("https://www.slideshare.net/slideshow/embed_code/key/aDfLeASoUQYFx0"));
        p_alkitab.setOnClickListener(v -> handlePdfClick("https://www.slideshare.net/slideshow/embed_code/key/F09kT2IosMpVwf"));
        p_aiSquared.setOnClickListener(v -> handlePdfClick("https://www.slideshare.net/slideshow/embed_code/key/MufsoIHKBIlR6V"));
        p_fokus.setOnClickListener(v -> handlePdfClick("https://www.slideshare.net/slideshow/embed_code/key/7GNdvGTo9M4Fc9"));
        p_bibleGpt.setOnClickListener(v -> handlePdfClick("https://www.slideshare.net/slideshow/embed_code/key/j402m0w1jgBb9m"));

        coba.setOnClickListener(view -> {
            Intent intent = new Intent(this, MainActivity.class);
            intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);
            startActivity(intent);
        });
    }

    private void showImage(int imageResId) {
        Dialog dialog = new Dialog(this);
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
        dialog.setContentView(R.layout.image_dialog);
        Objects.requireNonNull(dialog.getWindow()).setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);

        PhotoView photoView = dialog.findViewById(R.id.photoView);
        photoView.setImageResource(imageResId);

        ImageButton closeButton = dialog.findViewById(R.id.close);
        closeButton.setOnClickListener(v -> dialog.dismiss());

        dialog.setCancelable(true);
        dialog.setCanceledOnTouchOutside(true);
        dialog.getWindow().getDecorView().getRootView().setOnClickListener(v -> dialog.dismiss());
        dialog.show();
    }

    private void handleClick(String url, String title){
        if (isConnected){
            NetworkUtil.openWebView(this, url, title);
        } else {
            ToastUtil.showToast(this);
        }
    }

    private void handleYoutubeClick(String videoId, int startTime){
        if (isConnected){
            showYoutube(videoId, startTime);
        } else {
            ToastUtil.showToast(this);
        }
    }

    private void showYoutube(String videoId, int startTime) {
        Intent intent = new Intent(this, YoutubePlayer.class);
        intent.putExtra(YoutubePlayer.EXTRA_VIDEO_ID, videoId);
        intent.putExtra(YoutubePlayer.EXTRA_START_TIME, startTime);
        startActivity(intent);
    }

    private void handlePdfClick(String pdfUrl){
        if (isConnected){
            showPdf(pdfUrl);
        } else {
            ToastUtil.showToast(this);
        }
    }

    private void showPdf(String pdfUrl) {
        Intent intent = new Intent(this, materi.class);
        intent.putExtra(materi.EXTRA_URL, pdfUrl);
        startActivity(intent);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        NetworkUtil.unregisterNetworkChangeReceiver(this);
    }

    private void updateConnectionStatus(boolean isConnected) {
        this.isConnected = isConnected;

        Button[] buttons = {aidan, aisquared, fokus, alkitabGPT, v_aidan, v_alkitab, v_aiSquared, v_fokus, v_bibleGpt, p_aidan, p_alkitab, p_aiSquared, p_fokus, p_bibleGpt};

        for (Button button : buttons) {
            button.setEnabled(isConnected);
        }

        if (!isConnected){
            ToastUtil.showToast(this);
        }

    }

    @Override
    public void onNetworkChange(boolean isConnected) {
        updateConnectionStatus(isConnected);
    }

}

File: androidApp/src/main/java/org/sabda/gpt/Selengkapnya.kt
package org.sabda.gpt

import android.app.Dialog
import android.content.Intent
import android.os.Bundle
import android.view.ViewGroup
import android.view.Window
import android.widget.ImageButton
import androidx.appcompat.app.AppCompatActivity
import com.github.chrisbanes.photoview.PhotoView
import org.sabda.gpt.databinding.ActivitySelengkapnyaBinding
import org.sabda.gpt.utility.NetworkUtil
import org.sabda.gpt.utility.NetworkUtil.NetworkChangeCallback
import org.sabda.gpt.utility.showToast
import java.util.Objects

class Selengkapnya : AppCompatActivity(), NetworkChangeCallback {

    private lateinit var binding: ActivitySelengkapnyaBinding
    private var isConnected: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivitySelengkapnyaBinding.inflate(layoutInflater)
        setContentView(binding.root)

        isConnected = NetworkUtil.isNetworkAvailable(this)
        NetworkUtil.registerNetworkChangeReceiver(this) {isConnected ->
            this.updateConnectionStatus(isConnected)
        }

        updateConnectionStatus(isConnected)
        setupListener()
    }

    private fun setupListener() {
        binding.kembali.setOnClickListener { finish() }

        val imageViewMap = mapOf(
            binding.aidanAlkitab to R.drawable.aidanalkitab,
            binding.AISquareSquare to R.drawable.aisquare,
            binding.FOKUS to R.drawable.fokus,
            binding.AlkitabGPT to R.drawable.alkitabgptdetail,
            binding.bibleGpt to R.drawable.biblegpt
        )

        imageViewMap.forEach { (imageView, imageResId) ->
            imageView.setImageResource(imageResId)
        }

        // Image click listeners
        val imageButtonMap = mapOf(
            binding.aidanAlkitab to R.drawable.aidanalkitab,
            binding.AISquareSquare to R.drawable.aisquare,
            binding.FOKUS to R.drawable.fokus,
            binding.AlkitabGPT to R.drawable.alkitabgptdetail,
            binding.bibleGpt to R.drawable.biblegpt
        )

        imageButtonMap.forEach { (button, imageResId) ->
            button.setOnClickListener { showImage(imageResId) }
        }

        // Button click listeners
        val webButtonMap = mapOf(
            binding.aidanAlkitabButton to Pair("ai-dan-alkitab/", "Seminar: AI dan Alkitab"),
            binding.aisquaredbutton to Pair("pa-dengan-ai-cara-ai-squared/", "Metode AI Squared"),
            binding.fokusbutton to Pair("bagaimana-ai/", "Metode F.0.K.U.S"),
            binding.alkitabGPTbutton to Pair("alkitab-gpt/", "Seminar: Alkitab GPT")
        )

        webButtonMap.forEach { (button, data) ->
            button.setOnClickListener { handleClick("https://ai.sabda.org/${data.first}", data.second) }
        }

        binding.cobabutton.setOnClickListener {
            val intent = Intent(this, MainActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
            startActivity(intent)
        }

        // Video buttons
        val videoButtonMap = mapOf(
            binding.videoaidanAlkitab to Pair("gDLTyyabCvI", 218),
            binding.videoAlkitabGPT to Pair("Pf9DPEcX8JA", 797),
            binding.videoaiSquared to Pair("qHPwPjuAEs4", 783),
            binding.videofokus to Pair("GHcRwNXdnvQ", 0),
            binding.videoBibleGpt to Pair("vZkGKgtRPeU", 0)
        )

        videoButtonMap.forEach { (button, data) ->
            button.setOnClickListener { handleYoutubeClick(data.first, data.second) }
        }

        // PDF buttons
        val pdfButtonMap = mapOf(
            binding.pdfaidanAlkitab to "aDfLeASoUQYFx0",
            binding.pdfAlkitabGPT to "F09kT2IosMpVwf",
            binding.pdfaiSquared to "MufsoIHKBIlR6V",
            binding.pdffokus to "7GNdvGTo9M4Fc9",
            binding.pdfBibleGpt to "j402m0w1jgBb9m"
        )

//        pdfButtonMap.forEach { (button, url) ->
//            button.setOnClickListener { handlePdfClick("https://www.slideshare.net/slideshow/embed_code/key/${url}") }
//        }
    }

    private fun showImage(imageResId: Int) {
        val dialog = Dialog(this)
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE)
        dialog.setContentView(R.layout.image_dialog)
        Objects.requireNonNull(dialog.window)
            ?.setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)

        val photoView = dialog.findViewById<PhotoView>(R.id.photoView)
        photoView.setImageResource(imageResId)

        val closeButton = dialog.findViewById<ImageButton>(R.id.close)
        closeButton.setOnClickListener { dialog.dismiss() }

        dialog.setCancelable(true)
        dialog.setCanceledOnTouchOutside(true)
        dialog.window?.decorView?.rootView?.setOnClickListener { dialog.dismiss() }
        dialog.show()
    }

    private fun handleClick(url: String, title: String) {
        if (isConnected) NetworkUtil.openWebView(this, url, title) else applicationContext.showToast(getString(R.string.toast_offline))
    }

    private fun handleYoutubeClick(videoId: String, startTime: Int) {
        if (isConnected) showYoutube(videoId, startTime) else applicationContext.showToast(getString(R.string.toast_offline))
    }

    private fun showYoutube(videoId: String, startTime: Int) {
        val intent = Intent(this, YoutubePlayer::class.java)
        intent.putExtra(YoutubePlayer.EXTRA_VIDEO_ID, videoId)
        intent.putExtra(YoutubePlayer.EXTRA_START_TIME, startTime)
        startActivity(intent)
    }

//    private fun handlePdfClick(pdfUrl: String) {
//        if (isConnected) {
//            showPdf(pdfUrl)
//        } else {
//            ToastUtil.showToast(this)
//        }
//    }

//    private fun showPdf(pdfUrl: String) {
//        val intent = Intent(this, Materi::class.java)
//        intent.putExtra(Materi.EXTRA_URL, pdfUrl)
//        startActivity(intent)
//    }

    override fun onDestroy() {
        super.onDestroy()
        NetworkUtil.unregisterNetworkChangeReceiver(this)
    }

    private fun updateConnectionStatus(isConnected: Boolean) {
        this.isConnected = isConnected

        val buttons = listOf(
            binding.aidanAlkitabButton,
            binding.aisquaredbutton,
            binding.fokusbutton,
            binding.alkitabGPTbutton,
            binding.videoaidanAlkitab,
            binding.videoAlkitabGPT,
            binding.videoaiSquared,
            binding.videofokus,
            binding.videoBibleGpt,
            binding.pdfaidanAlkitab,
            binding.pdfAlkitabGPT,
            binding.pdfaiSquared,
            binding.pdffokus,
            binding.pdfBibleGpt
        )

        buttons.forEach { it.isEnabled = isConnected }

        if (!isConnected) applicationContext.showToast(getString(R.string.toast_offline))
    }

    override fun onNetworkChange(isConnected: Boolean) {
        updateConnectionStatus(isConnected)
    }
}
